algorithms/baselines/curious_ppo/agent_curious_ppo.py

```diff
@@ -0,0 +1,211 @@
+import copy
+import time
+from functools import partial
+
+import tensorflow as tf
+
+from algorithms.algo_utils import num_env_steps, maybe_extract_key, main_observation, goal_observation
+from algorithms.baselines.ppo.agent_ppo import AgentPPO, PPOBuffer
+from algorithms.curiosity.curiosity_model import CuriosityModel
+from algorithms.env_wrappers import main_observation_space
+from algorithms.tf_utils import placeholder_from_space, merge_summaries
+from utils.timing import Timing
+from utils.utils import AttrDict
+
+
+class CuriousPPOBuffer(PPOBuffer):
+    def __init__(self):
+        super(CuriousPPOBuffer, self).__init__()
+        self.next_obs = None
+
+    def reset(self):
+        super(CuriousPPOBuffer, self).reset()
+        self.next_obs = []
+
+    # noinspection PyMethodOverriding
+    def add(self, obs, next_obs, actions, action_probs, rewards, dones, values, goals=None):
+        """Append one-step data to the current batch of observations."""
+        args = copy.copy(locals())
+        super(CuriousPPOBuffer, self)._add_args(args)
+
+
+class AgentCuriousPPO(AgentPPO):
+    """PPO with a curiosity module (ICM or RND)"""
+    class Params(AgentPPO.Params):
+        """Hyperparams for curious PPO"""
+        def __init__(self, experiment_name):
+            super(AgentCuriousPPO.Params, self).__init__(experiment_name)
+
+            # add more params here  #TODO: Aleksei give defaults
+            self.cm_beta = 0.5
+            self.cm_lr_scale = 10.0
+            self.clip_bonus = 0.1
+            self.prediction_bonus_coeff = 0.05  # scaling factor for prediction bonus vs env rewards
+
+            self.stack_past_frames = 3
+            self.forward_fc = 512
+
+        @staticmethod
+        def filename_prefix():
+            return 'curious_ppo_'
+
+    def __init__(self, make_env_func, params):
+        super(AgentCuriousPPO, self).__init__(make_env_func, params)
+
+        env = self.make_env_func()  # we need it to query observation shape, number of actions, etc.
+        self.ph_next_observations = placeholder_from_space(main_observation_space(env))
+        self.prediction_curiosity_bonus = None  # updated in self.add_cm_objectives()
+
+        # Create graph for curiosity module (ICM)
+        self._cm = CuriosityModel(
+            env, self.ph_observations, self.ph_next_observations, self.ph_actions, params.stack_past_frames,
+            params.forward_fc, params=params)
+
+        # add ICM loss keys to objective function
+        self.objectives.update(self.add_cm_objectives())  # TODO: will this merging overwrite some keys?
```

Comment:
Good point, I think it's better to have a separate variable, like self.curiosity_objectives, just to be safe. But up to you,